### PR TITLE
Fix leaking taints in case of cloud provider error on node deletion

### PR DESCRIPTION
**Pull request by aleksandra-malinowska**
In case when node was successfully drained, but cloud provider error prevented its deletion, ToBeDeleted taint remains on the node. It happened in this test run:

https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gci-gce-autoscaling-migs/6052

### CODE REVIEW

File: cluster-autoscaler/core/scale_down.go [hunk 1]
```diff
@@ -638,11 +638,33 @@ func (sd *ScaleDown) waitForEmptyNodesDeleted(emptyNodes []*apiv1.Node, confirma
 }
 
 func deleteNode(context *AutoscalingContext, node *apiv1.Node, pods []*apiv1.Pod) errors.AutoscalerError {
```
Comment (mwielgus) [inline cluster-autoscaler/core/scale_down.go]: unit test please

File: cluster-autoscaler/core/scale_down.go [hunk 2]
```diff
@@ -638,11 +638,33 @@ func (sd *ScaleDown) waitForEmptyNodesDeleted(emptyNodes []*apiv1.Node, confirma
...
+	if err := deletetaint.MarkToBeDeleted(node, context.ClientSet); err != nil {
+		context.Recorder.Eventf(node, apiv1.EventTypeWarning, "ScaleDownFailed", "failed to mark the node as toBeDeleted/unschedulable: %v", err)
+		return errors.ToAutoscalerError(errors.ApiCallError, err)
+	}
+
+	// If we fail to evict all the pods from the node we want to remove delete taint
+	defer func() {
+		if !deleteSuccessful {
+			deletetaint.CleanToBeDeleted(node, context.ClientSet)
+			context.Recorder.Eventf(node, apiv1.EventTypeWarning, "ScaleDownFailed", "failed to drain the node, aborting ScaleDown")
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/scale_down.go]: this message is no longer correct, can you update it?

Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: I'm considering splitting it into two events (to better communicate point of failure), what do you think?

File: cluster-autoscaler/core/scale_down.go (hunk 1 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: sure:)

File: cluster-autoscaler/core/scale_down.go (hunk 2 above)
Comment (MaciekPytel) [reply cluster-autoscaler/core/scale_down.go]: I'm fine with both one or 2 different events, whichever makes more sense to you

File: cluster-autoscaler/core/scale_down.go (hunk 1 above)
Comment (aleksandra-malinowska) [reply cluster-autoscaler/core/scale_down.go]: done

File: cluster-autoscaler/core/scale_down.go [hunk 3]
```diff
@@ -638,11 +638,42 @@ func (sd *ScaleDown) waitForEmptyNodesDeleted(emptyNodes []*apiv1.Node, confirma
...
+	}
+
+	// If we fail to evict all the pods from the node we want to remove delete taint
+	defer func() {
+		if !deleteSuccessful {
+			deletetaint.CleanToBeDeleted(node, context.ClientSet)
+			if !drainSuccessful {
+				context.Recorder.Eventf(node, apiv1.EventTypeWarning, "ScaleDownFailed", "failed to drain the node, aborting ScaleDown")
+			} else {
+				context.Recorder.Eventf(node, apiv1.EventTypeWarning, "ScaleDownFailed", "failed to delete the node, aborting ScaleDown")
```
Comment (MaciekPytel) [inline cluster-autoscaler/core/scale_down.go]: I would argue that we're no longer aborting scale down at this point, we just failed it. Feel free to ignore this comment though.

File: cluster-autoscaler/core/scale_down_test.go [hunk 4]
```diff
@@ -334,9 +334,169 @@ func TestFindUnneededNodePool(t *testing.T) {
...
+		expectedDeletion  bool
+	}{
+		{
+			name:              "successful attempt to delete node with pods",
+			pods:              []string{"p1", "p2"},
+			drainSuccess:      true,
+			nodeDeleteSuccess: true,
+			expectedDeletion:  true,
+		},
+		/* Temporarily disabled as it takes several minutes due to hardcoded timeout.
```
Comment (aleksandra-malinowska) [inline cluster-autoscaler/core/scale_down_test.go]: Leaving this commented out for now as discussed offline with @mwielgus. We're too close to release to refactor now and this scenario increases the time unit tests for this package take from ~1s to ~2 min.

**Pull request merged**